lineoperations/src/lineoperations.c

```diff
@@ -0,0 +1,205 @@
+{
+	GeanyDocument *doc = NULL;
+	doc = document_get_current();
+
+	if(doc) sortlines(doc, 1);
+}
+
+
+/* Sort Lines Descending */
+static void
+action_sortdesc_item(GtkMenuItem *menuitem, gpointer gdata)
+{
+	GeanyDocument *doc = NULL;
+	doc = document_get_current();
+
+	if(doc) sortlines(doc, 0);
+}
+
+
+static gboolean
+lo_init(GeanyPlugin *plugin, gpointer gdata)
+{
+	GeanyData *geany_data = plugin->geany_data;
+
+
+	GtkWidget *submenu;
+	GtkWidget *sep1;
+	GtkWidget *sep2;
+	GtkWidget *rmdupst_item;
+	GtkWidget *rmdupln_item;
+	GtkWidget *rmunqln_item;
+	GtkWidget *rmemtyln_item;
+	GtkWidget *rmwhspln_item;
+	GtkWidget *sortasc_item;
+	GtkWidget *sortdesc_item;
+
+	main_menu_item = gtk_menu_item_new_with_mnemonic(_("_Line Operations"));
+	gtk_widget_show(main_menu_item);
+
+	submenu = gtk_menu_new();
+	gtk_widget_show(submenu);
+	sep1 = gtk_separator_menu_item_new();
+	sep2 = gtk_separator_menu_item_new();
+	rmdupst_item  = gtk_menu_item_new_with_mnemonic(_("Remove Duplicate Lines, _Sorted"));
+	rmdupln_item  = gtk_menu_item_new_with_mnemonic(_("Remove Duplicate Lines, _Ordered"));
+	rmunqln_item  = gtk_menu_item_new_with_mnemonic(_("Remove _Unique Lines"));
+	rmemtyln_item = gtk_menu_item_new_with_mnemonic(_("Remove _Empty Lines"));
+	rmwhspln_item = gtk_menu_item_new_with_mnemonic(_("Remove _Whitespace Lines"));
+	sortasc_item  = gtk_menu_item_new_with_mnemonic(_("Sort Lines _Ascending"));
+	sortdesc_item = gtk_menu_item_new_with_mnemonic(_("Sort Lines _Descending"));
+
+	gtk_widget_show(sep1);
+	gtk_widget_show(sep2);
+	gtk_widget_show(rmdupst_item);
+	gtk_widget_show(rmdupln_item);
+	gtk_widget_show(rmunqln_item);
+	gtk_widget_show(rmemtyln_item);
+	gtk_widget_show(rmwhspln_item);
+	gtk_widget_show(sortasc_item);
+	gtk_widget_show(sortdesc_item);
+
+	gtk_menu_shell_append(GTK_MENU_SHELL(submenu), rmdupst_item);
+	gtk_menu_shell_append(GTK_MENU_SHELL(submenu), rmdupln_item);
+	gtk_menu_shell_append(GTK_MENU_SHELL(submenu), rmunqln_item);
+	gtk_menu_shell_append(GTK_MENU_SHELL(submenu), sep1);
+	gtk_menu_shell_append(GTK_MENU_SHELL(submenu), rmemtyln_item);
+	gtk_menu_shell_append(GTK_MENU_SHELL(submenu), rmwhspln_item);
+	gtk_menu_shell_append(GTK_MENU_SHELL(submenu), sep2);
+	gtk_menu_shell_append(GTK_MENU_SHELL(submenu), sortasc_item);
+	gtk_menu_shell_append(GTK_MENU_SHELL(submenu), sortdesc_item);
+
+	gtk_menu_item_set_submenu(GTK_MENU_ITEM(main_menu_item), submenu);
+
+	gtk_container_add(GTK_CONTAINER(geany->main_widgets->tools_menu), main_menu_item);
+
+
+	g_signal_connect(rmdupst_item, "activate", G_CALLBACK(action_rmdupst_item), NULL);
+	g_signal_connect(rmdupln_item, "activate", G_CALLBACK(action_rmdupln_item), NULL);
+	g_signal_connect(rmunqln_item, "activate", G_CALLBACK(action_rmunqln_item), NULL);
+	g_signal_connect(rmemtyln_item, "activate", G_CALLBACK(action_rmemtyln_item), NULL);
+	g_signal_connect(rmwhspln_item, "activate", G_CALLBACK(action_rmwhspln_item), NULL);
```

Comment:
you could lower duplication of the wrapper callbacks by having a single callback that calls the appropriate function (either passed as parameter [1] or referenced as an ID or something), as they all have they all behave the same (but the sort ones).
You may or may not want this, it's merely a suggestion.  And it's even a bad one if you expect to alter the various functions to receive different parameters in the future.

[1] technically in plain C you're not supposed to convert a function pointer to a data pointer.  However, GTK depends on this and it effectively works on all supported platforms (if not all platforms in general).